### ISSUE TITLE
训练流程

### ISSUE THREAD

**Issue by con-tinue** [rubbish001/Co-Fix3d#13]
大佬，你的cofix3d两阶段训练中，第一阶段不训练LGE模块吗，是不是第一阶段用config里的LCv1，先跳过WavePaint，第二阶段再开始用WavePaint进行训练

**Comment by WenxuanLi-whu** [comment]
第一阶段训练第一个LGE模块，后面6轮训练剩下的2个LGE模块

**Comment by Da1symeeting1** [comment]
> 大佬，你的cofix3d两阶段训练中，第一阶段不训练LGE模块吗，是不是第一阶段用config里的LCv1，先跳过WavePaint，第二阶段再开始用WavePaint进行训练

哥们，请问你复现出来了吗，我跑出来好像没这么高的点，融合阶段的 config 有问题，我自己修改的跑完之后没 mAP 和 NDS 没达到。

**Comment by WenxuanLi-whu** [comment]
第一阶段的训练没有复现出来吗，第二阶段引用第一阶段的权重，你看看权重是否匹配上了，你看看第二阶段的训练日志，其中网络模型的部分梯度打开的部分是否匹配上了

**Comment by Da1symeeting1** [comment]
> 第一阶段的训练没有复现出来吗，第二阶段引用第一阶段的权重，你看看权重是否匹配上了，你看看第二阶段的训练日志，其中网络模型的部分梯度打开的部分是否匹配上了
我这边跑了 Lidar 的 v1和v2，然后LC的v1 和v2融合， Lidar的结果是MAP 66.4 LC的v1结果是 MAP 32多，然后LC v2的结果是mAP: 0.6906
mATE: 0.2901
mASE: 0.2493
mAOE: 0.2449
mAVE: 0.2128
mAAE: 0.1822
NDS: 0.7274，梯度打开的是 pts_bbox_head.inpaint_bev以及 pts_bbox_head的内容

**Comment by WenxuanLi-whu** [comment]
第一阶段纯点云的mAP差不多是66.4%左右，第二阶段训练结果为67.3%左右， 对于图像融合模态， 第一阶段直接用66.4的权重，然后第二阶段训练图像20轮（这个版本是43%左右，改进后的co-Fusuion是54%左右，具体的我也忘记了），然后冻结图像和点云分支的主干和LSS部分权重，在训练6轮 （CBGS），最后差不多就是论文的结果

**Comment by Da1symeeting1** [comment]
mAP: 0.6549
mATE: 0.2891
mASE: 0.2562
mAOE: 0.2447
mAVE: 0.2254
mAAE: 0.1892
NDS: 0.7070
Eval time: 94.4s 这是 Co-Fix3d_L_v1 的结果（未更改配置文件）20ep
mAP: 0.6631
mATE: 0.2911
mASE: 0.2541
mAOE: 0.2572
mAVE: 0.2074
mAAE: 0.1871
NDS: 0.7119
Eval time: 118.7s 这是 Co-Fix3d_L_v2 的结果（未更改配置文件）6ep
mAP: 0.3248
mATE: 0.7198
mASE: 0.2775
mAOE: 0.5195
mAVE: 0.8731
mAAE: 0.2465
NDS: 0.3988
Eval time: 112.9s 这是 Co-Fix3d_LC_r50_v1 的结果 20ep ，不知道你那个 43 是什么，这里使用的 step = 2，不是纯图像分支的检测吗，然后最后融合的时候两个都冻结掉了
mAP: 0.6906
mATE: 0.2901
mASE: 0.2493
mAOE: 0.2449
mAVE: 0.2128
mAAE: 0.1822
NDS: 0.7274 这是 Co-Fix3d_LC_r50_v2的结果（不更改配置文件无法跑，仅修改 cbgs），用的是 CBGS 10ep，lidar 用的是 L_v2 的权重，图像用的是 LC_r50_v1 的权重然后都冻结掉了。
使用的设备是 4 卡 4090

**Comment by WenxuanLi-whu** [comment]
请对比一下Co-Fix3D | L | 67.3 | 72.0 的日志的网络结构，仔细对比

**Comment by con-tinue** [comment]
佬，想问一下怎么合并训练集和验证集呀，我直接用
ann_file=
[
    data_root + 'nuscenes_infos_train.pkl',
    data_root + 'nuscenes_infos_val.pkl',
]
会报错说没有set_epoch,我看有人说合并train和val的.pkl文件，但是我合并完发现新的.PKL文件还没有train.pkl大，想问一下你是怎么合并进行训练的，如果是第一种，能不能讲讲是怎么解决报错的，如果是第二种，能不能给一份代码，感谢感谢

**Comment by WenxuanLi-whu** [comment]
合并train和val，或者你这种写法都可以，你这种写法要改代码

**Comment by Da1symeeting1** [comment]
> 佬，想问一下怎么合并训练集和验证集呀，我直接用 ann_file= [ data_root + 'nuscenes_infos_train.pkl', data_root + 'nuscenes_infos_val.pkl', ] 会报错说没有set_epoch,我看有人说合并train和val的.pkl文件，但是我合并完发现新的.PKL文件还没有train.pkl大，想问一下你是怎么合并进行训练的，如果是第一种，能不能讲讲是怎么解决报错的，如果是第二种，能不能给一份代码，感谢感谢

我之前试用过 bevfusion-adlab 用的是

<img width="473" height="417" alt="Image" src="https://github.com/user-attachments/assets/cd3d8b35-1595-4b64-a9ff-513992768b97" />

这种形式，并不需要更改代码什么的，不知道在mmdetection3d 中怎么样，这样的方式不行吗？我之前试过网上那种二合一的形式，这是变小了。

**Comment by WenxuanLi-whu** [comment]
调试一遍，跟踪程序一下就知道了

**Comment by con-tinue** [comment]
佬，想问一下你用的哪种方法呀

**Comment by WenxuanLi-whu** [comment]
我是重新生成pkl的方法，挂机一下就可以生成出来

**Comment by Da1symeeting1** [comment]
> 佬，想问一下你用的哪种方法呀

佬，成功之后可否也发我一份呢 👍

**Comment by con-tinue** [comment]
有代码吗佬，救一下

**Comment by con-tinue** [comment]
我不知道咋直接回复你da1佬，上面你说的直接变成列表会被默认变成concatdataset，那个里面没有set_epoch，会报错，我是这样

**Comment by con-tinue** [comment]
> > 佬，想问一下怎么合并训练集和验证集呀，我直接用 ann_file= [ data_root + 'nuscenes_infos_train.pkl', data_root + 'nuscenes_infos_val.pkl', ] 会报错说没有set_epoch,我看有人说合并train和val的.pkl文件，但是我合并完发现新的.PKL文件还没有train.pkl大，想问一下你是怎么合并进行训练的，如果是第一种，能不能讲讲是怎么解决报错的，如果是第二种，能不能给一份代码，感谢感谢
> 
> 我之前试用过 bevfusion-adlab 用的是
> 
> <img alt="Image" width="473" height="417" src="https://private-user-images.githubusercontent.com/70803671/510552055-cd3d8b35-1595-4b64-a9ff-513992768b97.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI0MDgwMjYsIm5iZiI6MTc2MjQwNzcyNiwicGF0aCI6Ii83MDgwMzY3MS81MTA1NTIwNTUtY2QzZDhiMzUtMTU5NS00YjY0LWE5ZmYtNTEzOTkyNzY4Yjk3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTA2VDA1NDIwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA4NDEwMmE2Y2RlODc0Y2QxMDAyNmQxNWM1MWZhNDU0NmUzZjhlZGVlY2QwNzdkYTIwMGFjNjRkZDkyMjU0M2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.zHMUXGf08SoiXSTAuqDrneh-GWpAAek1_qWGMZ9mC6I">
> 这种形式，并不需要更改代码什么的，不知道在mmdetection3d 中怎么样，这样的方式不行吗？我之前试过网上那种二合一的形式，这是变小了。

哦哦我会了，你试过的变小了的train+val.pkl跑过测试吗，精度正常不

**Comment by Da1symeeting1** [comment]
> > > 佬，想问一下怎么合并训练集和验证集呀，我直接用 ann_file= [ data_root + 'nuscenes_infos_train.pkl', data_root + 'nuscenes_infos_val.pkl', ] 会报错说没有set_epoch,我看有人说合并train和val的.pkl文件，但是我合并完发现新的.PKL文件还没有train.pkl大，想问一下你是怎么合并进行训练的，如果是第一种，能不能讲讲是怎么解决报错的，如果是第二种，能不能给一份代码，感谢感谢
> > 
> > 
> > 我之前试用过 bevfusion-adlab 用的是
> > <img alt="Image" width="473" height="417" src="https://private-user-images.githubusercontent.com/70803671/510552055-cd3d8b35-1595-4b64-a9ff-513992768b97.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI0MDgwMjYsIm5iZiI6MTc2MjQwNzcyNiwicGF0aCI6Ii83MDgwMzY3MS81MTA1NTIwNTUtY2QzZDhiMzUtMTU5NS00YjY0LWE5ZmYtNTEzOTkyNzY4Yjk3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTA2VDA1NDIwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA4NDEwMmE2Y2RlODc0Y2QxMDAyNmQxNWM1MWZhNDU0NmUzZjhlZGVlY2QwNzdkYTIwMGFjNjRkZDkyMjU0M2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.zHMUXGf08SoiXSTAuqDrneh-GWpAAek1_qWGMZ9mC6I">
> > 这种形式，并不需要更改代码什么的，不知道在mmdetection3d 中怎么样，这样的方式不行吗？我之前试过网上那种二合一的形式，这是变小了。
> 
> 哦哦我会了，你试过的变小了的train+val.pkl跑过测试吗，精度正常不

11/06 14:02:18 - mmengine - INFO - load 62964 traffic_cone database infos in DataBaseSampler
11/06 14:02:18 - mmengine - INFO - load 65262 truck database infos in DataBaseSampler
11/06 14:02:18 - mmengine - INFO - load 339949 car database infos in DataBaseSampler
11/06 14:02:18 - mmengine - INFO - load 161928 pedestrian database infos in DataBaseSampler
11/06 14:02:18 - mmengine - INFO - load 133804 barrier database infos in DataBaseSampler
11/06 14:02:18 - mmengine - INFO - load 11050 construction_vehicle database infos in DataBaseSampler
11/06 14:02:18 - mmengine - INFO - load 8846 motorcycle database infos in DataBaseSampler
11/06 14:02:18 - mmengine - INFO - load 8185 bicycle database infos in DataBaseSampler
11/06 14:02:18 - mmengine - INFO - load 12286 bus database infos in DataBaseSampler
11/06 14:02:18 - mmengine - INFO - load 19202 trailer database infos in DataBaseSampler
11/06 14:02:20 - mmengine - INFO - After filter database:
11/06 14:02:20 - mmengine - INFO - load 56328 traffic_cone database infos in DataBaseSampler
11/06 14:02:20 - mmengine - INFO - load 60675 truck database infos in DataBaseSampler
11/06 14:02:20 - mmengine - INFO - load 296294 car database infos in DataBaseSampler
11/06 14:02:20 - mmengine - INFO - load 149183 pedestrian database infos in DataBaseSampler
11/06 14:02:20 - mmengine - INFO - load 127275 barrier database infos in DataBaseSampler
11/06 14:02:20 - mmengine - INFO - load 10623 construction_vehicle database infos in DataBaseSampler
11/06 14:02:20 - mmengine - INFO - load 8090 motorcycle database infos in DataBaseSampler
11/06 14:02:20 - mmengine - INFO - load 7550 bicycle database infos in DataBaseSampler
11/06 14:02:20 - mmengine - INFO - load 11662 bus database infos in DataBaseSampler
11/06 14:02:20 - mmengine - INFO - load 18160 trailer database infos in DataBaseSampler
11/06 14:03:51 - mmengine - INFO - ------------------------------
11/06 14:03:51 - mmengine - INFO - The length of the dataset: 28130
11/06 14:03:51 - mmengine - INFO - The number of instances per category in the dataset:
+----------------------+--------+
| category             | number |
+----------------------+--------+
| car                  | 413318 |
| truck                | 72815  |
| construction_vehicle | 11993  |
| bus                  | 13163  |
| trailer              | 20701  |
| barrier              | 125095 |
| motorcycle           | 10109  |
| bicycle              | 9478   |
| pedestrian           | 185847 |
| traffic_cone         | 82362  |
+----------------------+--------+
11/06 14:05:01 - mmengine - INFO - load 62964 traffic_cone database infos in DataBaseSampler
11/06 14:05:01 - mmengine - INFO - load 65262 truck database infos in DataBaseSampler
11/06 14:05:01 - mmengine - INFO - load 339949 car database infos in DataBaseSampler
11/06 14:05:01 - mmengine - INFO - load 161928 pedestrian database infos in DataBaseSampler
11/06 14:05:01 - mmengine - INFO - load 133804 barrier database infos in DataBaseSampler
11/06 14:05:01 - mmengine - INFO - load 11050 construction_vehicle database infos in DataBaseSampler
11/06 14:05:01 - mmengine - INFO - load 8846 motorcycle database infos in DataBaseSampler
11/06 14:05:01 - mmengine - INFO - load 8185 bicycle database infos in DataBaseSampler
11/06 14:05:01 - mmengine - INFO - load 12286 bus database infos in DataBaseSampler
11/06 14:05:01 - mmengine - INFO - load 19202 trailer database infos in DataBaseSampler
11/06 14:05:03 - mmengine - INFO - After filter database:
11/06 14:05:03 - mmengine - INFO - load 56328 traffic_cone database infos in DataBaseSampler
11/06 14:05:03 - mmengine - INFO - load 60675 truck database infos in DataBaseSampler
11/06 14:05:03 - mmengine - INFO - load 296294 car database infos in DataBaseSampler
11/06 14:05:03 - mmengine - INFO - load 149183 pedestrian database infos in DataBaseSampler
11/06 14:05:03 - mmengine - INFO - load 127275 barrier database infos in DataBaseSampler
11/06 14:05:03 - mmengine - INFO - load 10623 construction_vehicle database infos in DataBaseSampler
11/06 14:05:03 - mmengine - INFO - load 8090 motorcycle database infos in DataBaseSampler
11/06 14:05:03 - mmengine - INFO - load 7550 bicycle database infos in DataBaseSampler
11/06 14:05:03 - mmengine - INFO - load 11662 bus database infos in DataBaseSampler
11/06 14:05:03 - mmengine - INFO - load 18160 trailer database infos in DataBaseSampler
11/06 14:05:18 - mmengine - INFO - ------------------------------
11/06 14:05:18 - mmengine - INFO - The length of the dataset: 6019
11/06 14:05:18 - mmengine - INFO - The number of instances per category in the dataset:
+----------------------+--------+
| category             | number |
+----------------------+--------+
| car                  | 80004  |
| truck                | 15704  |
| construction_vehicle | 2678   |
| bus                  | 3158   |
| trailer              | 4159   |
| barrier              | 26992  |
| motorcycle           | 2508   |
| bicycle              | 2381   |
| pedestrian           | 34347  |
| traffic_cone         | 15597  |
+----------------------+--------+ 刚试了一下，没合并也可以

**Comment by Da1symeeting1** [comment]
train_dataset_config = dict(
    type=dataset_type,
    data_root=data_root,
    ann_file='nuscenes_infos_train.pkl',
    pipeline=train_pipeline,
    metainfo=metainfo,
    modality=input_modality,
    test_mode=False,
    data_prefix=data_prefix,
    use_valid_flag=True,
    serialize_data=False,
    box_type_3d='LiDAR')

val_dataset_config = dict(
    type=dataset_type,
    data_root=data_root,
    ann_file='nuscenes_infos_val.pkl',
    pipeline=train_pipeline,
    metainfo=metainfo,
    modality=input_modality,
    test_mode=False,
    data_prefix=data_prefix,
    use_valid_flag=True,
    serialize_data=False,
    box_type_3d='LiDAR')

cbgs_train_config = dict(
    type='CBGSDataset',
    dataset=train_dataset_config
)

cbgs_val_config = dict(
    type='CBGSDataset',
    dataset=val_dataset_config
)

train_dataloader = dict(
    _delete_=True,
    batch_size=4,
    num_workers=4,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=True),
    dataset=dict(
        type='ConcatDataset',
        datasets=[cbgs_train_config, cbgs_val_config]
    )
)

val_dataloader = dict(
    _delete_=True,
    batch_size=2,
    num_workers=4,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        ann_file='nuscenes_infos_val.pkl',
        pipeline=test_pipeline,
        metainfo=metainfo,
        modality=input_modality,
        data_prefix=data_prefix,
        test_mode=True,
        box_type_3d='LiDAR',
        backend_args=backend_args))

test_dataloader = dict(
    _delete_=True,
    batch_size=1,
    num_workers=4,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        ann_file='nuscenes_infos_val.pkl',
        pipeline=test_pipeline,
        metainfo=metainfo,
        modality=input_modality,
        data_prefix=data_prefix,
        test_mode=True,
        box_type_3d='LiDAR',
        backend_args=backend_args))

**Comment by con-tinue** [comment]
天塌了，为什么我看nuscenes数据集的测试集挑战赛关闭了啊，就是说没办法提交测试集结果进行评估了，佬们有消息吗？
